skills/repo.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package skills
+
+import (
+	"errors"
+	"sort"
+	"sync"
+)
+
+// Repository stores all known snappy skills and slots and types.
+type Repository struct {
+	// Protects the internals from concurrent access.
+	m     sync.Mutex
+	types []Type
+}
+
+var (
+	// ErrDuplicate is reported when type, skill or slot already exist.
+	ErrDuplicate = errors.New("duplicate found")
+)
+
+// NewRepository creates an empty skill repository.
+func NewRepository() *Repository {
+	return &Repository{}
+}
+
+// AllTypes returns all skill types.
+func (r *Repository) AllTypes() []Type {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	types := make([]Type, len(r.types))
+	copy(types, r.types)
+	return types
```

Comment:
`return append([]Type(nil), r.types...)`

... but it actually feels somewhat like an early optimization. It's unclear what the use cases for this are. Might not even be worth exposing it for now.